updater/src/updater.rs

```diff
@@ -248,189 +301,406 @@ impl Updater {
-						Ok(b) => {
-							s.backoff = None;
-							b
-						},
-						Err(e) => {
-							let mut n = s.backoff.map(|b| b.0 + 1).unwrap_or(1);
-							s.backoff = Some((n, Instant::now() + Duration::from_secs(2usize.pow(n) as u64)));
-
-							return Err((format!("Unable to fetch update ({}): {:?}", fetched.version, e), false));
-						},
+		self.update_policy.path.join(name)
+	}
+
+	fn on_fetch(&self, latest: &OperationsInfo, res: Result<PathBuf, fetch::Error>) {
+		let mut state = self.state.lock();
+
+		// Bail out if the latest release has changed in the meantime
+		if state.latest.as_ref() != Some(&latest) {
+			return;
+		}
+
+		// The updated status should be set to fetching
+		if let UpdaterStatus::Fetching { ref release, binary, retries } = state.status.clone() {
+			match res {
+				// We've successfully fetched the binary
+				Ok(path) => {
+					let setup = |path: &Path| -> Result<(), String> {
+						let dest = self.updates_path(&Updater::update_file_name(&release.version));
+						if !dest.exists() {
+							info!(target: "updater", "Fetched latest version ({}) OK to {}", release.version, path.display());
+							fs::create_dir_all(dest.parent().expect("at least one thing pushed; qed")).map_err(|e| format!("Unable to create updates path: {:?}", e))?;
+							fs::copy(path, &dest).map_err(|e| format!("Unable to copy update: {:?}", e))?;
+							restrict_permissions_owner(&dest, false, true).map_err(|e| format!("Unable to update permissions: {}", e))?;
+							info!(target: "updater", "Copied updated binary to {}", dest.display());
+						}
+
+						Ok(())
 					};
 
-					info!(target: "updater", "Fetched latest version ({}) OK to {}", fetched.version, b.display());
-					fs::create_dir_all(dest.parent().expect("at least one thing pushed; qed")).map_err(|e| (format!("Unable to create updates path: {:?}", e), true))?;
-					fs::copy(&b, &dest).map_err(|e| (format!("Unable to copy update: {:?}", e), true))?;
-					restrict_permissions_owner(&dest, false, true).map_err(|e| (format!("Unable to update permissions: {}", e), true))?;
-					info!(target: "updater", "Installed updated binary to {}", dest.display());
-				}
-				let auto = match self.update_policy.filter {
-					UpdateFilter::All => true,
-					UpdateFilter::Critical if fetched.is_critical /* TODO: or is on a bad fork */ => true,
-					_ => false,
+					// There was a fatal error setting up the update, disable the updater
+					if let Err(err) = setup(&path) {
+						state.status = UpdaterStatus::Disabled;
+						warn!("{}", err);
+					} else {
+						state.status = UpdaterStatus::Ready { release: release.clone() };
+						// will lock self.state
+						drop(state);
+						self.updater_step();
+					}
+				},
+				// There was an error fetching the update, apply a backoff delay before retrying
+				Err(err) => {
+					let delay = 2usize.pow(retries) as u64;
+					// cap maximum backoff to 1 month
+					let delay = cmp::min(delay, 30 * 24 * 60 * 60);
```

Comment:
I'd say a day/week would be enough. To reach 2^21 (month) we already need to spend 48 days waiting for previous delays.

With week (2^19) it's 12 days waiting before reaching the cap
and with a day (2^16) it's 2 days waiting.

(Given my math is not off :weary: )

And I think it shouldn't be an issue to handle one request a week/day from every single node in case something goes wrong.

Reply:
#hedidthemath